plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1375,15 +1376,39 @@ public synchronized void cleanupQuery(ConnectorSession session)
             return;
         }
 
-        commit();
-
-        clearCurrentTransaction();
-        long transactionId = transaction.get().getTransactionId();
-        ScheduledFuture<?> heartbeatTask = transaction.get().getHeartbeatTask();
-        heartbeatTask.cancel(true);
-
-        // Any failure around aborted transactions, etc would be handled by Hive Metastore commit and TrinoException will be thrown
-        delegate.commitTransaction(identity, transactionId);
+        Throwable failure = null;
+        try {
+            commit();
+        }
+        catch (Throwable commitFailure) {
+            failure = commitFailure;
+        }
+        finally {
+            // we need to kill heartbeat thread and commit hive transaction to avoid transaction leak in HMS
+            try {
+                clearCurrentTransaction();
+                long transactionId = transaction.get().getTransactionId();
+                ScheduledFuture<?> heartbeatTask = transaction.get().getHeartbeatTask();
+                heartbeatTask.cancel(true);
+
+                // Any failure around aborted transactions, etc would be handled by Hive Metastore commit and TrinoException will be thrown
+                delegate.commitTransaction(identity, transactionId);
```

Comment:
It seems weird to commit transaction if our commit failed (ie when `failure != null`)

Reply:
Yeah - I agree - though I am not sure if just changing this code to explicit `abortTransaction` here would be good too. Is there a chance that our `commit` did already partial work and it is too later for abort in HMS?

Reply:
fortunately there is no "legacy behavior" -- we didn't commit/abort tx when our commit failed, so we can do whatever we want

i didn't test it, but abort seems way more appropriate... if we choose to call commit for some reason, we need to document why we think it's a good idea.